src/program/types/CallExpression.js

```diff
@@ -57,9 +57,20 @@ export default class CallExpression extends Node {
 					if (this.callee.object.type === 'Identifier') {
 						context = this.callee.object.name;
 					} else {
-						context = this.findScope(true).createDeclaration('ref');
 						const callExpression = this.callee.object;
-						code.prependRight(callExpression.start, `(${context} = `);
+
+						// Prepend the declaraction in place to prevent an expression
+						// begining with a parenthesis. Other types are guarded by
+						// an expression before the parenthesis so the declaration
+						// is created at the top of the scope.
+						if (this.parent.type === 'ExpressionStatement') {
```

Comment:
This check is not enough if the `CallExpression` is not the expression of the `ExpressionStatement`, like this:

```js
const foo = { bar: [] }
const baz = true ? [1,2,3] : []
foo.bar.push(...baz)()
```

Reply:
Good point. A better check here would be if the start of parent is also the start of the statement.

Reply:
I tried the following a few days ago:

```diff
                                        } else {
                                                context = this.findScope(true).createDeclaration('ref');
                                                const callExpression = this.callee.object;
-                                           code.prependRight(callExpression.start, `(${context} = `);
+                                         const statement = this.findNearest(/Statement/);
+                                         code.prependRight(callExpression.start, (statement.start === callExpression.start ? ';' : '') + `(${context} = `);
                                                code.appendLeft(callExpression.end, `)`);
                                        }
```

That broke the following test, though:

```js
function foo (x) {
        if ( x )
          return ref => (bar || baz).Test( ref, ...arguments );
      }
```